### DISCUSSION TITLE
How to generate shared wasm file with `dylink.0` section

### DISCUSSION THREAD

**Discussion by ApsarasX** [WebAssembly/binaryen#4615] [Q&A]
Clang and Emscripten both could generate shared/PIC wasm file using the following commands.

```bash
clang --target=wasm32-unknown-emscripten -nostdlib -Wl,--shared -fPIC demo.c  -o demo.wasm
```
```bash
emcc demo.c -Os -s WASM=1 -s SIDE_MODULE=1 -o demo.wasm
```

So how does binaryen generate shared wasm file with `dylink.0` section?

**Comment by kripken** [comment]
Binaryen support for the dylink section is here: (this is the binary writing part)

https://github.com/WebAssembly/binaryen/blob/b337a11b0d2ac2af1b8c0241854e600be480792a/src/wasm/wasm-binary.cpp#L1126

I believe how it all works together is that passing `SIDE_MODULE` to emcc will make emcc call `wasm-ld` with `--experimental-pic -shared`. Then a dylink section is emitted. Binaryen work afterwards (wasm-emscripten-finalize, wasm-opt) will then read it and write it.

**Reply by ApsarasX** [reply]
Does Binaryen have to rely on emcc to emit a dylink section? 

Can't Binaryen itself emit a dylink section?

Sorry, maybe I didn't make my application scenario clear.

I am using AssemblyScript based on Binaryen to generate wasm.

And I need to generate a main wasm module and a side wasm module(as dynamic link library, just like `.so` file) both running on my custom wasm runtime.

I'm looking for some methods in Binaryen that have the same effect as `-experimental-pic -shared`, but so far no luck.

**Reply by kripken** [reply]
Internally that is possible, basically this data structure should be filled out:

https://github.com/WebAssembly/binaryen/blob/0b0581a2db2ea0f28e820957504e431881cb1f70/src/wasm.h#L1964-L1965

However AssemblyScript uses Binaryen through binaryen.js, and I don't think we have a JS or C API for that yet. That would need to be added first, I think.
